src/pybamm/config.py

```diff
@@ -45,20 +48,125 @@ def is_running_tests():  # pragma: no cover
     return False
 
 
-def ask_user_opt_in(timeout=10):
+def get_input_or_timeout(timeout):  # pragma: no cover
     """
-    Ask the user if they want to opt in to telemetry.
+    Cross-platform input with timeout, using various methods depending on the
+    environment. Works in Jupyter notebooks, Windows, and Unix-like systems.
+
+    Args:
+        timeout (float): Timeout in seconds
+
+    Returns:
+        str | None: The user input if received before the timeout, or None if:
+            - The timeout was reached
+            - Telemetry is disabled via environment variable
+            - Running in a non-interactive environment
+            - An error occurred
+
+        The caller can distinguish between an empty response (user just pressed Enter)
+        which returns '', and a timeout/error which returns None.
+    """
+    # Check for telemetry disable flag
+    if os.getenv("PYBAMM_DISABLE_TELEMETRY", "false").lower() != "false":
+        return None
+
+    if not (sys.stdin.isatty() or is_notebook()):
+        return None
 
-    Parameters
-    ----------
-    timeout : float, optional
-        The timeout for the user to respond to the prompt. Default is 10 seconds.
+    # 1. Handling for Jupyter notebooks. This is a simplified
+    # implementation in comparison to widgets because notebooks
+    # are usually run interactively, and we don't need to worry
+    # about the event loop. The input is disabled when running
+    # tests due to the environment variable set.
+    if is_notebook():  # pragma: no cover
+        try:
+            from IPython.display import clear_output
+
+            user_input = input("Do you want to enable telemetry? (Y/n): ")
+            clear_output()
+
+            return user_input
+
+        except Exception:  # pragma: no cover
+            return None
+
+    # 2. Windows-specific handling
+    if sys.platform == "win32":
+        try:
+            import msvcrt
+
+            start_time = time.time()
+            input_chars = []
+            sys.stdout.write("Do you want to enable telemetry? (Y/n): ")
+            sys.stdout.flush()
+
+            while time.time() - start_time < timeout:
+                if msvcrt.kbhit():
+                    char = msvcrt.getwche()
+                    if char in ("\r", "\n"):
+                        sys.stdout.write("\n")
+                        return "".join(input_chars)
+                    input_chars.append(char)
+                time.sleep(0.1)
+            return None
+        except Exception:
+            return None
 
-    Returns
-    -------
-    bool
-        True if the user opts in, False otherwise.
+    # 3. POSIX-like systems will need to use termios
+    else:  # pragma: no cover
+        try:
+            import termios
+            import tty
+            import select
+
+            # Save terminal settings for later
+            old_settings = termios.tcgetattr(sys.stdin)
```

Comment:
It seems excessive that we need to change the terminal settings for this

Reply:
What this allows you to do is that you don't get new indentation when you write to the terminal again. I wrapped this in a try-finally block based on the Python docs: https://docs.python.org/3/library/termios.html#example.

So, by saving the attributes, I'm trying to disable the fact that `sys.stdout.write("\n")` will not return the cursor to the leftmost position (because we are in raw mode and we need to do that manually instead – for which I want to avoid).

We end up with this, otherwise:

```
❯ python -c "import pybamm; print(pybamm.IDAKLUSolver())"
PyBaMM can collect usage data and send it to the PyBaMM team to help us improve the software.
We do not collect any sensitive information such as models, parameters, or simulation results - only information on which parts of the code are being used and how frequently.
This is entirely optional and does not impact the functionality of PyBaMM.
For more information, see https://docs.pybamm.org/en/latest/source/user_guide/index.html#telemetry
Do you want to enable telemetry? (Y/n): n

                                         Telemetry disabled.
                                                            <pybamm.solvers.idaklu_solver.IDAKLUSolver object at 0x102de3ad0>
                                                                                                                             %
```

and this would break one's terminal until one restarts it – a more robust solution could be to use `TCSAFLUSH` instead of `TCSADRAIN`, I'll switch to that.